### PR TITLE
stdio dialer support

### PR DESCRIPTION
**Pull request by aluzzardi**
- engine: stdio dialer support (e.g. `dagger dial-stdio` command)
- Go SDK: dial stdio support

**NOTE**: Go SDK **still** uses the embedded engine. the stdio dialer is for testing purposes, the real user are going to be the TS/Python SDKs.

In action:

```console
$ dagger do -f git.graphql
{
    "core": {
        "git": {
        }
    }
}

$ DAGGER_HOST=bin:///Users/al/.bin/dagger dagger do -f git.graphql
{
    "core": {
        "git": {
        }
    }
}
```

/cc @helderco

### CODE REVIEW

**Comment by helderco** [comment]
>  [...] the real user are going to be the TS/Python SDKs

Sorry, I'm missing some context.

- Is `dagger dial-stdio` to replace `dagger dev` when running an (embedded) client?
- Does anything change for extensions?
- When I run `go run ./cmd/dagger dial-stdio` I get the following panic:

```
http: panic serving : too many concurrent operations on a single file or socket (max 1048575)
goroutine 1051810 [running]:
net/http.(*conn).serve.func1()
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/http/server.go:1850 +0xb0
panic({0x105a47e00, 0x105be5dc0})
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/runtime/panic.go:890 +0x258
internal/poll.(*fdMutex).rwlock(0x140001ae000, 0x80?)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/internal/poll/fd_mutex.go:147 +0x134
internal/poll.(*FD).readLock(...)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/internal/poll/fd_mutex.go:221
internal/poll.(*FD).Read(0x140001ae000, {0x14470b3f000, 0x1000, 0x1000})
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/internal/poll/fd_unix.go:144 +0x44
os.(*File).read(...)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/os/file_posix.go:31
os.(*File).Read(0x140001ac000, {0x14470b3f000?, 0x1434dc94690?, 0x1446ebf7e38?})
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/os/file.go:118 +0x5c
main.(*stdioConn).Read(0x1446ebf7e30?, {0x14470b3f000?, 0x16318ea00?, 0x16318e900?})
	/Users/helder/Projects/dagger/cmd/dagger/dial_stdio.go:66 +0x2c
net/http.(*connReader).Read(0x1446ebf7e30, {0x14470b3f000, 0x1000, 0x1000})
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/http/server.go:786 +0x234
bufio.(*Reader).fill(0x144704aa8a0)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/bufio/bufio.go:106 +0xfc
bufio.(*Reader).ReadSlice(0x144704aa8a0, 0x7f?)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/bufio/bufio.go:372 +0x30
bufio.(*Reader).ReadLine(0x144704aa8a0)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/bufio/bufio.go:401 +0x24
net/textproto.(*Reader).readLineSlice(0x1446ebf7e60)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/textproto/reader.go:56 +0x78
net/textproto.(*Reader).ReadLine(...)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/textproto/reader.go:37
net/http.readRequest(0xc0cb6eaef7664040?)
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/http/request.go:1036 +0x54
net/http.(*conn).readRequest(0x1445c160f00, {0x105bf54c8, 0x1446e5ffc40})
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/http/server.go:994 +0x1e4
net/http.(*conn).serve(0x1445c160f00, {0x105bf5570, 0x14000a9e5d0})
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/http/server.go:1916 +0x2e4
created by net/http.(*Server).Serve
	/opt/homebrew/Cellar/go/1.19.1/libexec/src/net/http/server.go:3102 +0x444
```

**Comment by aluzzardi** [comment]
> Is dagger dial-stdio to replace dagger dev when running an (embedded) client?

Yep.

- `dagger dev` is a workaround: it's supposed to be called by humans, not code
- Even if we keep communication over tcp, it should be a different command (`dagger engine start`? I don't know)
- `dial-stdio` allows to communicate over stdio rather than tcp (avoid binding a port, port collision, etc)

> Does anything change for extensions?

Not really.

As part of `dagger.Connect()` (Go SDK), the SDK supports `DAGGER_HOST`:

- `bin://<path to dagger>` -- uses dial-stdio
- For extensions, `DAGGER_HOST` is automatically set to `unix:///dagger.sock` (same behavior as today)

> When I run go run ./cmd/dagger dial-stdio I get the following panic

Ouch. Yeah this PR is not ready

**Comment by aluzzardi** [comment]
This PR should be ready to go

@helderco @dolanor The Python & Node SDKs should be able to communicate over stdio as well

File: sdk/go/internal/engineconn/bin/bin.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"net"
+	"net/http"
+	"net/url"
+
+	"dagger.io/dagger/internal/engineconn"
+	"github.com/docker/cli/cli/connhelper/commandconn"
+)
+
+func init() {
+	engineconn.Register("bin", New)
```
Comment (dolanor) [inline sdk/go/internal/engineconn/bin/bin.go:15]: Why `bin`?
Why not `stdio`? (is that what would be the defined scheme? (xxx://)

File: sdk/go/internal/engineconn/bin/bin.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+var _ engineconn.EngineConn = &Bin{}
+
+type Bin struct {
+	path string
+}
+
+func New(u *url.URL) (engineconn.EngineConn, error) {
+	return &Bin{
```
Comment (dolanor) [inline sdk/go/internal/engineconn/bin/bin.go:25]: Shouldn't we check that the scheme is valid here as well?

File: cmd/dagger/dial_stdio.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+func (l *stdioConnListener) Accept() (net.Conn, error) {
+	var conn net.Conn
+	l.once.Do(func() {
+		conn = &stdioConn{
+			closeCh: l.closeCh,
+		}
+	})
+	// Return stdio connection ONLY on the first `Accept()`
+	// since we can't multiplex on stdio.
```
Comment (sipsma) [inline cmd/dagger/dial_stdio.go:76]: Are there any immediate needs for multiplexing? I think there may be in the future w/ services and/or filesync stuff, just wondering if I'm not thinking of already existing needs.

File: cmd/dagger/main.go [hunk 4]
```diff
@@ -38,6 +39,8 @@ func init() {
 	devCmd.Flags().IntVar(&devServerPort, "port", 8080, "dev server port")
 	devCmd.Flags().StringSliceVarP(&localDirsInput, "local-dir", "l", []string{}, "local directory to import")
 
+	dialStdioCmd.Flags().StringSliceVarP(&localDirsInput, "local-dir", "l", []string{}, "local directory to import")
```
Comment (sipsma) [inline cmd/dagger/main.go:42]: no longer needed 🎉

File: sdk/go/internal/engineconn/bin/bin.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package embedded
```
Comment (sipsma) [inline sdk/go/internal/engineconn/bin/bin.go:1]: Should be `bin` right?

**Pull request closed without merging**